### PR TITLE
sys_heap_aligned_alloc() implementation

### PR DESCRIPTION
**Pull request by andyross**
Add the aligned_alloc() implementation that's been discussed in the past.  It works at the block level, returning an aligned memory region while returning the misaligned prefix/suffix regions to the heap.  You pay for the alignment with some extra fragmentation, but lose no memory.

### CODE REVIEW

**Review by dcpleung — APPROVED** [review]

**Comment by npitre** [comment]
Would it be possible for my other PR to be merged first?

It's been waiting for quite a while at this point and I've rebased it to 
account for new changes a few times already.

**Comment by andyross** [comment]
It still had some failures when I tried to cherry pick it in before this, unfortunately.  Really there's nothing blocking that PR, it just needs a passing CI.

**Comment by npitre** [comment]
On Fri, 19 Jun 2020, Andy Ross wrote:

> It still had some failures when I tried to cherry pick it in before this, unfortunately.  Really there's nothing blocking that PR, it just needs a passing CI.

What again? Everything passed last time I touched it.

And frankly speaking, failures were rather stupid given that tests are 
too much relying on behaviors of the previous allocator. To make one 
test that relies on under-allocation requires the compat layer to 
compensate which then makes tests relying on over-allocation to fail.

Do we really need to maintain this whack-a-mole game?

**Comment by carlescufi** [comment]
@npitre I assume you are referring to #24249. I have re-triggered the test that failed there, I think it was a spurious failure.

**Review by ceolin — APPROVED** [review]

**Comment by andyross** [comment]
Rebased over the optimization series which just merged.

**Review by tyrantlizard77 — APPROVED** [review]
thanks!! once this and some stack stuff I have lands we'll be able to dynamically allocate thread stacks

**Comment by andyross** [comment]
Close/open to retrigger CI.  Seems like that run failed with some kind of network burp ("ERROR: update failed for project openthread")

File: lib/os/heap.c [hunk 1]
```diff
@@ -9,6 +9,10 @@
 
 static void *chunk_mem(struct z_heap *h, chunkid_t c)
 {
+	if (c == 0) {
+		return NULL;
+	}
+
```
Comment (finikorg) [inline lib/os/heap.c:15]: is it OK to declare variable not in the beginning of the scope block?

**Review by finikorg — APPROVED** [review]
with small comment

File: lib/os/heap.c (hunk 1 above)
Comment (andyross) [reply lib/os/heap.c:15]: No one complains when I do it, so I haven't stopped. :)

I don't know that we have an official policy on this.  Linux kernel style definitely forbids it, but our checkpatch variant doesn't warn about it (I don't know if upstream does and we disabled it, or if we're stale, or whatever).

But I do it all the time.  IMHO locally-declared variables are a big win for readability and maintainability.

**Comment by npitre** [comment]
I know this is merged already, but here's my review comment (better late 
than never).

```
void *sys_heap_aligned_alloc(struct sys_heap *heap, size_t align, size_t bytes)
{
        struct z_heap *h = heap->heap;

        CHECK((align & (align - 1)) == 0);
        CHECK(big_heap(h));
```

What if you have small heaps and still want aligned allocations for only some of them?
Perhaps this could be a flag provided to sys_heap_init() and stored in the header so that large chunk headers are forced when this flag is set. 

```
        /* Find a free block that is guaranteed to fit */
        size_t chunksz = bytes_to_chunksz(h, bytes);
        size_t mask = (align / CHUNK_UNIT) - 1;
        size_t padsz = MAX(CHUNK_UNIT, chunksz + mask);
```

Isn't this going to go terribly wrong if align < CHUNK_UNIT?
You'd end up with mask = 0xffffffff.

As long as the big_heap restriction is there, I'd suggest simply falling back to sys_heap_alloc() up front when align <= CHUNK_UNIT.

Also the MAX() statement is wrong. Here you're comparing CHUNK_UNIT (in bytes) against chunksz (in chunks). They're not the same measurement units so to speak.
Anyway, with the above suggestion you should be able to get rid of it.

```
        chunkid_t c0 = alloc_chunks(h, padsz);

        if (c0 == 0) {
                return NULL;
        }
```

Of course we know that the over-allocation may fail although some smaller chunk with proper alignment may exist.
Might be worth adding a comment to that effect.

```
        /* Align within memory, using "chunk index" units.  Remember
         * the block we're aligning starts in the chunk AFTER the
         * header!
         */
        size_t c0i = ((size_t) &chunk_buf(h)[c0 + 1]) / CHUNK_UNIT;
        size_t ci = ((c0i + mask) & ~mask);
        chunkid_t c = c0 + (ci - c0i);
```

Semantically speaking, those size_t should probably be uintptr_t here.

And I suspect most people not intimately familiar with the sys_heap code 
might find this part rather criptic. I might have an idea to simplify 
things here.

```
        /* Split and free unused prefix */
        if (c > c0) {
                split_chunks(h, c0, c);
                set_chunk_used(h, c, true);
                free_chunks(h, c0);
        }

        /* Split and free unused suffix */
        if (chunksz < chunk_size(h, c)) {
                split_chunks(h, c, c + chunksz);
                set_chunk_used(h, c, true);
                free_chunks(h, c + chunksz);
        }
```

Here it would be better to use free_list_add() rather than 
free_chunks(), at least for the first case. We know the prefix chunk 
can't be merged.

Finally, I don't get this solo_free_header() thing. I'm guessing that 
this is to deal with alignment that may cause the prefix chunk to become 
too small. But once free_list_add() has ignored such a chunk, I don't 
see how this chunk ever get salvaged afterwards. As to the suffix chunk: 
this one should be split only if large enough, just like we do in 
split_alloc(). Maybe the best solution here would consist in a free list 
of unalocatable chunks waiting to be merged backinto larger chunks.

**Pull request merged**